dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+
+-- | This module contains the implementation of the @dhall tags@ command
+
+module Dhall.ETags
+    ( -- * ETags
+      generate
+    ) where
+
+import Control.Exception (handle, SomeException(..))
+import Data.List (isSuffixOf)
+import Data.Maybe (fromMaybe)
+import Data.Semigroup (Semigroup(..))
+import Dhall.Map (foldMapWithKey)
+import Data.Text (Text)
+import Data.Text.Encoding (encodeUtf8)
+import Dhall.Util (Input(..))
+import Dhall.Core (Expr(..), Binding(..))
+import Dhall.Src (Src(srcStart))
+import Dhall.Parser (exprFromText)
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
+
+import qualified Data.ByteString as BS (length)
+import qualified Data.Map      as M
+import qualified Data.Text     as T
+import qualified Data.Text.IO  as TIO
+import qualified System.Directory as SD
+
+{- 
+    Documentation for ETags format is not very informative and not very correct.
+    You can find some documentation here:
+    https://en.wikipedia.org/wiki/Ctags#Etags_2
+    and you can also check the source code here:
+    http://cvs.savannah.gnu.org/viewvc/vtags/vtags/vtags.el?view=markup
+-}
+
+data LineColumn = LC 
+    { lcLine :: Int
+      -- ^ line number, starting from 1, where to find the tag
+    , lcColumn :: Int
+      -- ^ column of line where tag is
+    } deriving (Eq, Ord, Show)
+
+data LineOffset = LO 
+    { loLine :: Int
+      -- ^ line number, starting from 1, where to find the tag
+    , loOffset :: Int
+      -- ^ byte offset from start of file. Not sure if any editor uses it
+    } deriving (Eq, Ord, Show)
+
+newtype Tags = Tags (M.Map FilePath [(LineOffset, Tag)])
+
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith (<>) ts1 ts2)
+
+instance Monoid Tags where
+    mempty = Tags M.empty
+    mappend = (<>)
+
+{-| For example, for the line: @let foo = \"foo\"@ the tag is:
+    > Tag "let " "foo"
+-}
+data Tag = Tag
+    { tagPattern :: Text
+      -- ^ In vtags source code this field is named \"pattern\" and EMacs used it as
+      --   a regex pattern to locate line with tag. It's looking for ^<tag pattern>.
+      --   Looks like vi is not using it.
+    , tagName :: Text
+      -- ^ text, that editor compare with selected text. So it's really name of entity
+    } deriving (Show)
+
+type LineNumber = Int
+
+type ByteOffset = Int
+
+{-| Generate ETags for Dhall expressions
+-}
+generate
+    :: Input
+    -- ^ Where to look for files. This can be a directory name (@.@ for example),
+    --   a file name or `StandardInput`. If `StandardInput`, then this will wait for
+    --   file names from @STDIN@.
+    --   This way someone can combine tools in @bash@ to send, for example, output from
+    --   @find@ to the input of @dhall tags@.
+    -> Maybe [Text]
+    -- ^ List of suffixes for dhall files or Nothing to check all files ([".dhall"] by default)
```

Comment:
Since the defaulting to `[".dhall"]` is handled in the options parser, I don't think we should mention it here.

```suggestion
    -- ^ List of suffixes for dhall files or Nothing to check all files
```

Reply:
Changed.